### ISSUE TITLE
QST:

### ISSUE THREAD

**Issue by BENR0** [geopandas/geopandas#2374]
- [x] I have searched the [geopandas] tag on [StackOverflow](https://stackoverflow.com/questions/tagged/geopandas) and [GIS StackExchange](https://gis.stackexchange.com/questions/tagged/geopandas) for similar questions.

---

#### Question about geopandas

This question has been asked and I have read the conversations in the respective issues as well as on stackoverflow. I get the 
`python3.8/site-packages/geopandas/_compat.py:111: UserWarning: The Shapely GEOS version (3.9.1-CAPI-1.14.2) is incompatible with the GEOS version PyGEOS was compiled with (3.8.1-CAPI-1.13.3). Conversions between both will be slow.` warning.

I installed shapely and pygeos via pip in a pyenv environment and used the `--no-binary` flag in order to use the system GEOS version which is 3.8.1 (provided by an environment module). `geopandas.show_versions()` reports version 3.8.1 and the correct GEOS lib path.

I don't know what I am missing. Any hint is appreciated.

**Comment by brendan-ward** [comment]
I've found this is sensitive to the order that you install things.  If you install geopandas first (via `pip install geopandas`), it will automatically install `shapely` from the wheel rather than building against your local GEOS, and then a subsequent call to `pip install shapely --no-binary shapely` doesn't appear to do anything because it doesn't fully uninstall them first.

I suggest `pip uninstall shapely pygeos` followed by `pip install shapely --no-binary shapely pygeos` just to be sure that these are uninstalled before building and reinstalling.  This works for me for this particular issue (MacOS w/ pyenv).

**Comment by BENR0** [comment]
@brendan-ward yes I had a hunch that something like this could be a factor which is why I uninstalled all three packages and reinstalled them as you described but to no avail. That's why I as a last resort open this issue ;-).

**Comment by brendan-ward** [comment]
Are you using a virtual environment as well?  Is it possible there is a mixup between packages installed into your base Python vs your specific Python environment?

**Comment by BENR0** [comment]
Yes I am using a virtual environment but without using system site packages. I also already checked for mixups too and that's not it either. Are there maybe any other packages which install GEOS as a binary?

**Comment by brendan-ward** [comment]
One thing to try would be to uninstall `shapely` and verify that it is indeed no longer present in your virtual environment and base environment; that would at least help detect if you are running into issues from having multiple copies installed.

**Comment by BENR0** [comment]
Ok so I think I have figured out what is going on. I think me getting the warning is a bug which might be very specific to the "setup" I am working with.

I have a GEOS version 3.4.2 on the system which I can "override" with a newer version 3.8.1 by loading an environment module.
I run Python (3.8.12) in a virtual environment without using system site packages. Furthermore:

- geopandas 0.10.2
- shapely 1.8.1.post1 (installed with `--no-binary` option)
- pygeos 0.12.0 (installed with `--no-binary` option)

As I said I get the "incompatible version" warning even though `geopandas.show_versions()` shows me the correct version (3.8.1) and path for the GEOS library.

So what is going on?

After loading the environment module with the newer GEOS version I installed shapely and pygeos with `--no-binary`. This uses `geos-config` to find the GEOS library to use. Because the environment module modifies the `PATH` variable the newer version is used.

Now on import the compatibility is checked based on the version reported by `shapely.geos.geos_version_string`: https://github.com/geopandas/geopandas/blob/ee8adfb27659e9f982ba8cdadbf62c6b36dcc053/geopandas/_compat.py#L105

While the version reported by `geopandas.show_versions()` is determined by `shapely._buildcfg.geos_version`:
https://github.com/geopandas/geopandas/blob/ee8adfb27659e9f982ba8cdadbf62c6b36dcc053/geopandas/tools/_show_versions.py#L48-L49

The difference is that the first is using the `find_library` function from the `ctypes.util` module which in turn uses `ldconfig` (https://docs.python.org/3/library/ctypes.html#finding-shared-libraries) to find the path to the library which reports the older GEOS version.

On the other hand `shapely._buildcfg.geos_version` at first also tries to use `geos-config` to get the version.
Therefore I get a version mismatch even though I think, as long as I have the environment module loaded, the correct (newer) version is used by geopandas.

As a sidenote; I think the reason for this is that `ldconfig` uses cached pathes and loading an environment module does not rerun `ldconfig`. This might be solved by setting `LD_LIBRARY_PATH` in the module config file (I can not check because I do not have access to the config).

Sorry for the lengthy comment. I admit this might be a rather special situation but I hope this explanation makes sense. It would be nice if someone can confirm that this is indeed the correct explanation.
